tritium-metrics/src/main/java/com/palantir/tritium/metrics/CacheMetricSet.java

```diff
@@ -19,113 +19,108 @@
+    static CacheMetricSet create(Cache<?, ?> cache, String cacheName) {
+        return new CacheMetricSet(cache, cacheName);
     }
 
     @Override
     public Map<String, Metric> getMetrics() {
-        ImmutableMap.Builder<String, Metric> cacheMetrics = ImmutableMap.builder();
-
-        cacheMetrics.put(cacheMetricName("estimated", "size"),
-                new CachedGauge<Long>(clock, 500, TimeUnit.MILLISECONDS) {
-                    @Override
-                    protected Long loadValue() {
-                        return cache.size();
-                    }
-                });
-
-        cacheMetrics.put(cacheMetricName("request", "count"),
-                derivedGauge(CacheStats::requestCount));
-
-        cacheMetrics.put(cacheMetricName("hit", "count"),
-                derivedGauge(CacheStats::hitCount));
-
-        cacheMetrics.put(cacheMetricName("hit", "ratio"),
-                derivedGauge(stats -> stats.hitCount() / (1.0d * stats.requestCount())));
-
-        cacheMetrics.put(cacheMetricName("miss", "count"),
-                derivedGauge(CacheStats::missCount));
-
-        cacheMetrics.put(cacheMetricName("miss", "ratio"),
-                derivedGauge(stats -> stats.missCount() / (1.0d * stats.requestCount())));
-
-        cacheMetrics.put(cacheMetricName("eviction", "count"),
-                derivedGauge(CacheStats::evictionCount));
-
-        cacheMetrics.put(cacheMetricName("load", "success", "count"),
-                derivedGauge(CacheStats::loadSuccessCount));
-
-        cacheMetrics.put(cacheMetricName("load", "failure", "count"),
-                derivedGauge(CacheStats::loadExceptionCount));
-
-        cacheMetrics.put(cacheMetricName("load", "average", "millis"),
-                derivedGauge(stats -> stats.averageLoadPenalty() / 1000000.0d));
-
-        return cacheMetrics.build();
+        return Collections.unmodifiableMap(InternalCacheMetrics.createMetrics(
+                GuavaStats.create(cache, 1, TimeUnit.SECONDS),
+                metricName -> cacheName + '.' + metricName));
     }
 
+    static class GuavaStats implements InternalCacheMetrics.Stats {
+        private final Cache<?, ?> cache;
+        private final Supplier<CacheStats> stats;
+
+        @VisibleForTesting
+        GuavaStats(Cache cache, Supplier<CacheStats> stats) {
+            this.cache = cache;
+            this.stats = stats;
+        }
+
+        static InternalCacheMetrics.Stats create(Cache cache, long duration, TimeUnit timeUnit) {
+            Supplier<CacheStats> statsSupplier = Suppliers.memoizeWithExpiration(cache::stats, duration, timeUnit);
+            return new GuavaStats(cache, statsSupplier);
+        }
+
+        private CacheStats stats() {
+            return stats.get();
+        }
+
+        @Override
+        public long estimatedSize() {
+            return cache.size();
+        }
+
+        @Override
+        public long weightedSize() {
```

Comment:
how can maximum and weighted size be -1 for a guava cache? It supports a maximum size. Is it just because the size is registered when it's built but not supplied on the interface?

Reply:
Yeah, right now there is not a good way to get max size/weight or weighted size. Could try a reflection based field hack.

Reply:
We could also make these couple optional and just not register them if they’re not available

Reply:
@carterkozak and others -- thoughts on removing `weightedSize` & `maximumSize` from all registered cache metrics? Currently we only produce them for Caffeine caches as Guava does not easily expose an equivalent, and not sure they're super useful anyway (see https://github.com/google/guava/issues/1035 ). 

If we really wanted, we could use something like the following fragile package access code (or reflection equivalent):
```
package com.google.common.cache;

import java.util.OptionalLong;

public final class GuavaCacheSpy {
    private GuavaCacheSpy() {}

    public static OptionalLong weightedSize(Cache<?, ?> cache) {
        if (cache instanceof LocalCache.LocalManualCache) {
            LocalCache localCache = ((LocalCache.LocalManualCache) cache).localCache;
            if (localCache != null) {
                long totalWeight = 0;
                for (LocalCache.Segment segment : localCache.segments) {
                    if (segment != null) {
                        totalWeight += segment.totalWeight;
                    }
                }
                return OptionalLong.of(totalWeight);
            }
        }
        return OptionalLong.empty();
    }

    public static OptionalLong maxWeight(Cache<?, ?> cache) {
        if (cache instanceof LocalCache.LocalManualCache) {
            LocalCache localCache = ((LocalCache.LocalManualCache) cache).localCache;
            if (localCache != null) {
                return OptionalLong.of(localCache.maxWeight);
            }
        }
        return OptionalLong.empty();
    }
}
```

Reply:
I have used neither `weightedSize` nor `maximumSize`, though I can't speak for everyone.

My preference is to make them optional, and report values when they're available, that way we avoid bundling potential breaks with this feature.
I'd like to avoid the package hack and reflective access unless there's a very strong need, and even then I'd prefer migrating usage from guava to caffeine.

Reply:
Going to let these `return -1` for Guava caches for now, clarified JavaDoc on the `InternalCacheMetrics.Stats` interface, and move folks to over to Caffeine caches where they're still using Guava. See https://github.com/palantir/tritium/pull/93 for historical context where maximum & weighted size were added to untagged metrics.